# Parte 0 - Tres fallas que no fallan

## 0.a — El PDF que se indexa vacío (3 %) 

**`ejemplos/instructivo_escaneado.pdf` no tiene capa de texto. Ingesta la carpeta rag_pipeline.py y pega la salida de la ingesta. Después, en `ejemplos/` con python ingestion.py, explica en dos frases qué comprueba load_corpus para no indexarlo y qué produciría sin esa comprobación con ese mismo archivo: un fragmento hecho de [page=1] [page=2], sin una palabra del documento y sin ninguna excepción.**

In [11]:
from pathlib import Path
import os
import subprocess
import sys
import subprocess
import csv
import pandas as pd
from dotenv import load_dotenv

In [2]:
raiz=Path.cwd()

In [3]:
ejemplos = raiz / "ejemplos"
pdf_escaneado = ejemplos / "instructivo_escaneado.pdf"

In [4]:
print("Directorio del proyecto:", raiz)
print("Carpeta ejemplos:", ejemplos)
print("PDF escaneado:", pdf_escaneado)

print("\nComprobaciones:")
print("ingestion.py:", (raiz / "ingestion.py").is_file())
print("rag_pipeline.py:", (raiz / "rag_pipeline.py").is_file())
print("carpeta ejemplos:", ejemplos.is_dir())
print("PDF escaneado:", pdf_escaneado.is_file())

Directorio del proyecto: c:\Users\ASUS\Desktop\Maestria IA USFQ\IA Generativa y Agentes\Semana2\Taller2\taller-02-rag-alvarez-ballesteros-ludena
Carpeta ejemplos: c:\Users\ASUS\Desktop\Maestria IA USFQ\IA Generativa y Agentes\Semana2\Taller2\taller-02-rag-alvarez-ballesteros-ludena\ejemplos
PDF escaneado: c:\Users\ASUS\Desktop\Maestria IA USFQ\IA Generativa y Agentes\Semana2\Taller2\taller-02-rag-alvarez-ballesteros-ludena\ejemplos\instructivo_escaneado.pdf

Comprobaciones:
ingestion.py: True
rag_pipeline.py: True
carpeta ejemplos: True
PDF escaneado: True


Se procede a ejecutar el archivo `rag_pipeline.py`.

In [5]:
os.environ["EMBEDDING_BACKEND"] = "local"
os.environ["QDRANT_URL"] = ":memory:"
os.environ["CORPUS_DIR"] = "ejemplos"

print("EMBEDDING_BACKEND =", os.environ["EMBEDDING_BACKEND"])
print("QDRANT_URL        =", os.environ["QDRANT_URL"])
print("CORPUS_DIR        =", os.environ["CORPUS_DIR"])

EMBEDDING_BACKEND = local
QDRANT_URL        = :memory:
CORPUS_DIR        = ejemplos


In [6]:
resultado = subprocess.run(
    [sys.executable, "rag_pipeline.py"],
    capture_output=True,
    text=True
)

print(resultado.stdout)

if resultado.stderr:
    print("STDERR:")
    print(resultado.stderr)

print("Código de retorno:", resultado.returncode)

embeddings: local · sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2 · tope 128 tokens · Qdrant :memory:
ingesta: nimbus_gastos.md: 1 página(s), 294 caracteres útiles → 1 fragmento(s)
ingesta: nimbus_remoto.md: 1 página(s), 296 caracteres útiles → 1 fragmento(s)
ingesta: nimbus_vacaciones.md: 1 página(s), 363 caracteres útiles → 1 fragmento(s)
indexados 3 fragmentos de 3 documento(s)
  0.123  nimbus_vacaciones.md / nimbus_vacaciones-0000  # Política de vacaciones de NimbusSoft Los empleados a tiempo completo acumulan …
  0.118  nimbus_remoto.md / nimbus_remoto-0000  # Política de trabajo remoto de NimbusSoft El trabajo remoto está permitido hast…
  0.078  nimbus_gastos.md / nimbus_gastos-0000  # Política de reembolso de gastos de NimbusSoft Los gastos de viaje se reembolsa…

[modo inspección (sin clave ni Ollama)] abstuvo=None
[modo inspección: no hay clave de API ni Ollama; este es el prompt]
Responde usando solo el contexto recuperado.
Si el cont

In [8]:
resultado.stdout

'embeddings: local · sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2 · tope 128 tokens · Qdrant :memory:\ningesta: nimbus_gastos.md: 1 página(s), 294 caracteres útiles → 1 fragmento(s)\ningesta: nimbus_remoto.md: 1 página(s), 296 caracteres útiles → 1 fragmento(s)\ningesta: nimbus_vacaciones.md: 1 página(s), 363 caracteres útiles → 1 fragmento(s)\nindexados 3 fragmentos de 3 documento(s)\n  0.123  nimbus_vacaciones.md / nimbus_vacaciones-0000  # PolÃ\xadtica de vacaciones de NimbusSoft Los empleados a tiempo completo acumulan …\n  0.118  nimbus_remoto.md / nimbus_remoto-0000  # PolÃ\xadtica de trabajo remoto de NimbusSoft El trabajo remoto está permitido hast…\n  0.078  nimbus_gastos.md / nimbus_gastos-0000  # PolÃ\xadtica de reembolso de gastos de NimbusSoft Los gastos de viaje se reembolsa…\n\n[modo inspección (sin clave ni Ollama)] abstuvo=None\n[modo inspección: no hay clave de API ni Ollama; este es el prompt]\nResponde usando solo el contexto r

In [12]:
# Crear carpeta de resultados
carpeta_resultados = Path("resultados/parte0")
carpeta_resultados.mkdir(parents=True, exist_ok=True)

# Guardar stdout en CSV
archivo_csv = carpeta_resultados / "salida_0a.csv"

with open(archivo_csv, "w", newline="", encoding="utf-8-sig") as f:
    writer = csv.writer(f)
    writer.writerow(["salida_stdout"])
    writer.writerow([resultado.stdout])

print(f"\nSalida guardada en: {archivo_csv}")


Salida guardada en: resultados\parte0\salida_0a.csv


La salida de la ingesta fue almacenada dentro de la carpeta `resultados` dentro del presente repositorio. Mientras que las dos frases concluyentes que comprueban la funcionalidad de `load_corpus` y la consistencia de su salida, se describen al final del script `ingestion.py`.

## 0.b El fragmento que se corta a 128 

**Con el MiniLM de la Parte 0 (EMBEDDING_BACKEND=local, tope de 128 tokens), ingesta ejemplos/ pidiendo chunk_tokens=900 y pega el aviso. Después toma un texto de unas 900 palabras, cuenta sus tokens con el tokenizador del modelo, y compara el vector del texto entero con el del texto recortado a 
max_seq_length tokens: el coseno es ≈ 1. Todo lo que no cupo nunca llegó al índice. Es el ejercicio 5.1 del lunes, sobre el laboratorio.**

In [20]:
from pathlib import Path
from rag_pipeline import RagPipeline
import os
import numpy as np

In [15]:


os.environ["EMBEDDING_BACKEND"] = "local"
os.environ["QDRANT_URL"] = ":memory:"
os.environ["CORPUS_DIR"] = "ejemplos"

print("EMBEDDING_BACKEND =", os.environ["EMBEDDING_BACKEND"])
print("QDRANT_URL        =", os.environ["QDRANT_URL"])

EMBEDDING_BACKEND = local
QDRANT_URL        = :memory:


In [16]:


pipeline = RagPipeline()

chunks_900 = pipeline.ingest(
    Path("ejemplos"),
    chunk_tokens=900
)

print(f"\nFragmentos obtenidos: {len(chunks_900)}")
print(f"Tope del modelo: {pipeline.max_seq_tokens} tokens")

c:\Users\ASUS\AppData\Local\Programs\Python\Python313\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
Loading weights: 100%|██████████| 199/199 [00:00<00:00, 1702.74it/s]


ingesta: nimbus_gastos.md: 1 página(s), 294 caracteres útiles → 1 fragmento(s)
ingesta: nimbus_remoto.md: 1 página(s), 296 caracteres útiles → 1 fragmento(s)
ingesta: nimbus_vacaciones.md: 1 página(s), 363 caracteres útiles → 1 fragmento(s)

Fragmentos obtenidos: 3
Tope del modelo: 128 tokens


AVISO ingesta: chunk_tokens=900 supera el tope del modelo (128 tokens): lo que pase de 126 tokens no llega al índice —el MiniLM lo trunca en silencio; la H200 rechaza la petición—.
AVISO ingesta: instructivo_escaneado.pdf: 0 caracteres útiles en 2 página(s). ¿Escaneado o fotografiado? NO se indexa: sin texto no hay nada que recuperar (OCR aparte).


In [17]:
faqs = [
    "Para resetear tu contraseña entra a Configuración > Seguridad y pulsa 'Olvidé mi clave'.",
    "Los reembolsos se procesan en 5 a 7 días hábiles tras aprobar la solicitud.",
    "Puedes exportar tus datos en formato CSV desde el panel de administración.",
    "La factura electrónica se emite automáticamente al confirmar el pago.",
    "Para cancelar tu suscripción ve a Cuenta > Plan y selecciona 'Cancelar'.",
    "El soporte técnico atiende de lunes a viernes de 9h a 18h por chat.",
    "La API permite hasta 100 solicitudes por minuto en el plan básico.",
    "Aceptamos tarjetas de crédito, débito y transferencias bancarias.",
]

# Texto de unas 900 palabras, como en el Ejercicio 5.1 del lunes
texto_largo = (" ".join(faqs) + " ") * 8

print("Palabras del texto:", len(texto_largo.split()))

Palabras del texto: 768


In [18]:
### se obtiene el tokenizador y el tope del modelo

modelo = pipeline.encoder._modelo

tokenizador = modelo.tokenizer
tope = modelo.max_seq_length

print("Modelo:", modelo)
print("max_seq_length:", tope)

Modelo: SentenceTransformer(
  (0): Transformer({'transformer_task': 'feature-extraction', 'modality_config': {'text': {'method': 'forward', 'method_output_name': 'last_hidden_state'}}, 'module_output_name': 'token_embeddings', 'architecture': 'BertModel'})
  (1): Pooling({'embedding_dimension': 384, 'pooling_mode': 'mean', 'include_prompt': True})
)
max_seq_length: 128


In [19]:
### se ceuntan los tokens y se construye el recorte

ids = tokenizador(
    texto_largo,
    add_special_tokens=True,
    truncation=False
)["input_ids"]

# Mismo criterio del Ejercicio 5.1 del profesor:
# se conservan los tokens que caben dentro del límite del modelo.
recorte = tokenizador.decode(ids[1:tope - 1])

print("Tokens del texto:", len(ids))
print("Tope del modelo:", tope)
print("Tokens del recorte:", len(
    tokenizador(
        recorte,
        add_special_tokens=True,
        truncation=False
    )["input_ids"]
))

[transformers] Token indices sequence length is longer than the specified maximum sequence length for this model (1106 > 128). Running this sequence through the model will result in indexing errors


Tokens del texto: 1106
Tope del modelo: 128
Tokens del recorte: 128


In [21]:
### se comparan los embeddings


v_todo, v_recorte = pipeline.encoder.encode(
    [texto_largo, recorte]
)

coseno = float(v_todo @ v_recorte)

print(f"Tokens del texto: {len(ids)}")
print(f"Max_seq_length del modelo: {tope}")
print(f"Coseno(texto entero, recortado a {tope} tokens) = {coseno:.4f}")


Tokens del texto: 1106
Max_seq_length del modelo: 128
Coseno(texto entero, recortado a 128 tokens) = 1.0000


Se guardan los resultados del ejercicio dentro de la carpeta `resultados/parte0` incluidos los tokens y el coseno.

In [22]:
carpeta_resultados = Path("resultados/parte0")
carpeta_resultados.mkdir(parents=True, exist_ok=True)

archivo = carpeta_resultados / "salida_0b.csv"

with open(archivo, "w", newline="", encoding="utf-8-sig") as f:
    writer = csv.writer(f)

    writer.writerow(["elemento", "valor"])
    writer.writerow(["tokens_del_texto", len(ids)])
    writer.writerow(["max_seq_length", tope])
    writer.writerow(["tokens_del_recorte", len(
        tokenizador(
            recorte,
            add_special_tokens=True,
            truncation=False
        )["input_ids"]
    )])
    writer.writerow(["coseno", coseno])
    writer.writerow(["chunk_tokens_prueba", 900])
    writer.writerow(["chunk_tokens_parte1", 512])

print(f"Resultados guardados en: {archivo}")

Resultados guardados en: resultados\parte0\salida_0b.csv


In [25]:
N = len(ids)
N

1106

In [23]:
coseno

1.0

### Conclusión


El texto de prueba contiene aproximadamente 900 palabras y el tokenizador
del MiniLM produjo [1106] tokens, mientras que el `max_seq_length` del modelo
es de 128 tokens. El coseno entre el embedding del texto completo y el
embedding del texto recortado a 128 tokens fue [COSENO=1], mostrando que el
modelo representa únicamente la parte de la entrada que cabe dentro de su
tope; el contenido restante no llega al embedding ni, por tanto, al índice.

Para la Parte 1 utilizaré `chunk_tokens=512`. Aunque bge-m3 permite hasta
8192 tokens, que un fragmento quepa en el límite del modelo no significa
que sea un buen fragmento para recuperación: un fragmento demasiado grande
puede mezclar demasiada información y reducir la precisión con la que un
vecino recuperado responde a una pregunta concreta.

## 0.c El índice que no se queja

**Sobre 
ejemplos/ indexado, pregunta algo cuya respuesta no está (
"¿cuál es la política de mascotas?"
) y
pega los tres vecinos con sus puntajes. Tienen aspecto sano. Después pregunta algo cuya respuesta sí está en el
PDF escaneado que la 0.a no indexó (
"¿qué exige el instructivo de prácticas de campo sobre el seguro
de accidentes?"
) y pega los tres vecinos. También tienen aspecto sano.**

In [26]:
import os

os.environ["EMBEDDING_BACKEND"] = "local"
os.environ["QDRANT_URL"] = ":memory:"
os.environ["CORPUS_DIR"] = "ejemplos"

print("EMBEDDING_BACKEND =", os.environ["EMBEDDING_BACKEND"])
print("QDRANT_URL =", os.environ["QDRANT_URL"])
print("CORPUS_DIR =", os.environ["CORPUS_DIR"])

EMBEDDING_BACKEND = local
QDRANT_URL = :memory:
CORPUS_DIR = ejemplos


In [27]:
from pathlib import Path
from rag_pipeline import RagPipeline

pipeline = RagPipeline()

chunks = pipeline.ingest(
    Path("ejemplos")
)

print(f"Fragmentos obtenidos: {len(chunks)}")
print(f"Tope del modelo: {pipeline.max_seq_tokens} tokens")

Loading weights: 100%|██████████| 199/199 [00:00<00:00, 2685.36it/s]


ingesta: nimbus_gastos.md: 1 página(s), 294 caracteres útiles → 1 fragmento(s)
ingesta: nimbus_remoto.md: 1 página(s), 296 caracteres útiles → 1 fragmento(s)
ingesta: nimbus_vacaciones.md: 1 página(s), 363 caracteres útiles → 1 fragmento(s)
Fragmentos obtenidos: 3
Tope del modelo: 128 tokens


AVISO ingesta: instructivo_escaneado.pdf: 0 caracteres útiles en 2 página(s). ¿Escaneado o fotografiado? NO se indexa: sin texto no hay nada que recuperar (OCR aparte).


In [28]:
## indexando a quadrant en memoria
pipeline.index(chunks)

print("Índice creado correctamente.")
print("Colección:", pipeline.collection)

Índice creado correctamente.
Colección: mmia6013_rag


In [29]:
pregunta_1 = "¿Cuál es la política de mascotas?"

resultados_1 = pipeline.retrieve(
    pregunta_1,
    top_k=3
)

print("PREGUNTA:")
print(pregunta_1)
print("\nTOP 3 VECINOS:\n")

for i, resultado in enumerate(resultados_1, start=1):
    print(f"--- Vecino {i} ---")
    print(f"Score:    {resultado['score']:.4f}")
    print(f"Documento: {resultado['document']}")
    print(f"Chunk ID:  {resultado['chunk_id']}")
    print(f"Texto:    {resultado['text']}")
    print()

PREGUNTA:
¿Cuál es la política de mascotas?

TOP 3 VECINOS:

--- Vecino 1 ---
Score:    0.0632
Documento: nimbus_gastos.md
Chunk ID:  nimbus_gastos-0000
Texto:    # Política de reembolso de gastos de NimbusSoft Los gastos de viaje se reembolsan presentando factura dentro de los 30 días posteriores al gasto. El límite diario de alimentación en viajes es de 45 USD. Los pasajes aéreos deben comprarse en clase económica salvo vuelos de más de 8 horas, donde se permite económica premium con aprobación del gerente de área.

--- Vecino 2 ---
Score:    0.0339
Documento: nimbus_remoto.md
Chunk ID:  nimbus_remoto-0000
Texto:    # Política de trabajo remoto de NimbusSoft El trabajo remoto está permitido hasta 3 días por semana para todos los roles excepto soporte de infraestructura on-site. Los días remotos se coordinan con el líder de equipo. Para trabajar desde el exterior del país se requiere aprobación de Recursos Humanos con 30 días de anticipación y un máximo de 60 días por año.

--- Vecino

In [30]:
pregunta_2 = (
    "¿Qué exige el instructivo de prácticas de campo "
    "sobre el seguro de accidentes?"
)

resultados_2 = pipeline.retrieve(
    pregunta_2,
    top_k=3
)

print("PREGUNTA:")
print(pregunta_2)
print("\nTOP 3 VECINOS:\n")

for i, resultado in enumerate(resultados_2, start=1):
    print(f"--- Vecino {i} ---")
    print(f"Score:    {resultado['score']:.4f}")
    print(f"Documento: {resultado['document']}")
    print(f"Chunk ID:  {resultado['chunk_id']}")
    print(f"Texto:    {resultado['text']}")
    print()

PREGUNTA:
¿Qué exige el instructivo de prácticas de campo sobre el seguro de accidentes?

TOP 3 VECINOS:

--- Vecino 1 ---
Score:    0.2289
Documento: nimbus_remoto.md
Chunk ID:  nimbus_remoto-0000
Texto:    # Política de trabajo remoto de NimbusSoft El trabajo remoto está permitido hasta 3 días por semana para todos los roles excepto soporte de infraestructura on-site. Los días remotos se coordinan con el líder de equipo. Para trabajar desde el exterior del país se requiere aprobación de Recursos Humanos con 30 días de anticipación y un máximo de 60 días por año.

--- Vecino 2 ---
Score:    0.1679
Documento: nimbus_gastos.md
Chunk ID:  nimbus_gastos-0000
Texto:    # Política de reembolso de gastos de NimbusSoft Los gastos de viaje se reembolsan presentando factura dentro de los 30 días posteriores al gasto. El límite diario de alimentación en viajes es de 45 USD. Los pasajes aéreos deben comprarse en clase económica salvo vuelos de más de 8 horas, donde se permite económica premium co

### Conclusiones

Hit Rate es una métrica que mide si el sistema de recuperación logró encontrar al menos un fragmento relevante dentro de los primeros \(k\) resultados.

Hit Rate no puede detectar estos casos porque mide si aparece un fragmento relevante dentro de los resultados, pero no determina si la pregunta tiene respuesta en el corpus; por eso una pregunta sin respuesta puede recibir vecinos plausibles y una pregunta cuya respuesta está en un documento no indexado puede recibir vecinos aunque la respuesta sea irrecuperable.